crates/red_knot_python_semantic/resources/mdtest/known_constants.md

```diff
@@ -26,21 +26,67 @@ from typing import TYPE_CHECKING as TC
 reveal_type(TC)  # revealed: Literal[True]
 ```
 
-### Must originate from `typing`
+### User-defined `TYPE_CHECKING`
 
-Make sure we only use our special handling for `typing.TYPE_CHECKING` and not for other constants
-with the same name:
+If we set `TYPE_CHECKING = False` directly instead of importing it from the `typing` module, it will
+still be treated as `True` during type checking. This behavior is for compatibility with other major
+type checkers, e.g. mypy and pyright.
+
+Only `False` is allowed to be assigned to `TYPE_CHECKING`, and any assignment other than `False`
+will result in an error.
 
 `constants.py`:
 
 ```py
-TYPE_CHECKING: bool = False
+TYPE_CHECKING = False
+
+# error: [invalid-type-checking-constant]
+TYPE_CHECKING = True
+# error: [invalid-type-checking-constant]
+TYPE_CHECKING: bool = True
+# error: [invalid-type-checking-constant]
+TYPE_CHECKING: int = 1
+# error: [invalid-assignment]
+TYPE_CHECKING: str = False
```

Comment:
I think it would be clearer if we separate these lines (and the explanation paragraph above) into a separate test, so we can keep this test focused on the correct usage, without the reader wondering if there might be some interaction between the erroring lines and the correct line.